source/extensions/filters/http/wait/filter.h

```diff
@@ -0,0 +1,54 @@
+#pragma once
+
+#include "envoy/server/filter_config.h"
+#include "api/envoy/config/filter/http/wait/v2/wait_filter.pb.validate.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Wait {
+
+class WaitingFilter : public Http::StreamDecoderFilter,
+                      Logger::Loggable<Logger::Id::filter>,
+                      public Http::UpstreamCallbacks {
+public: 
+  WaitingFilter();
+  ~WaitingFilter();
+
+  // Http::FunctionalFilterBase
+  void onDestroy() override {}
+
+  // Http::StreamDecoderFilter
+  Http::FilterHeadersStatus decodeHeaders(Http::RequestHeaderMap &,
+                                          bool) override;
+
+  // Do nothing
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override {
+    return Http::FilterDataStatus::Continue;
+  }
+
+  Http::FilterTrailersStatus decodeTrailers(Http::RequestTrailerMap&) override {
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+  void setDecoderFilterCallbacks(
+      Http::StreamDecoderFilterCallbacks &callbacks) override {
+    decoder_callbacks_ = &callbacks;
+    if (callbacks.upstreamCallbacks()) {
```

Comment:
should we check `decoder_callbacks_->upstreamCallbacks()->upstream()` is not nil here, and skip adding us to the callbacks?

Reply:
That's what this does, if upstreamCallbacks is set this is an upstream filter. Also at construction Upstream won't be available afaik

Reply:
even if we re-use an existing connection?